File: 890-find-and-replace-pattern/890-find-and-replace-pattern.c
bool match(char *word, char *pattern) {
    int i,j;
    int len;
    char *dict;

    len = strlen(pattern);
    if (len == 1) {
        return true;
    }
    
    dict = calloc(26, sizeof(char));
    if (dict == NULL) {
        return false;
    }
        
    for (i = 0; i < len; i++) {
        if (dict[pattern[i]-'a'] == 0) {
            for (j = 0; j < 26; j++) {
                if (dict[j] == word[i]) {
                    return false;
                }
            }
            dict[pattern[i]-'a'] = word[i];     
        } else {
            if (dict[pattern[i]-'a'] != word[i]) {
                return false;
            }
        }
    }
    
    return true;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
char ** findAndReplacePattern(char ** words, int wordsSize, char * pattern, int* returnSize){
    int i, idx;
    char **result;
    
    *returnSize = 0;
    result = calloc(wordsSize, sizeof(char *));
    if (result == NULL) {
        return NULL;
    }
    
    idx = 0;
    for (i = 0; i < wordsSize; i++) {
        if (match(words[i], pattern)) {
            result[idx++] = words[i];
        }
    }

    result = realloc(result, idx*sizeof(char *));
    if (result == NULL) {
        return NULL;
    }

    *returnSize = idx;
    return result;
}

File: 890-find-and-replace-pattern/890-find-and-replace-pattern.py
class Solution(object):
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        
        def is_matched(word, pattern):
            if len(pattern) == 1:
                return True
            
            d = {}
            
            for i in range(0,len(pattern)):
                if pattern[i] not in d:
                    if word[i] in d.values():
                        return False
                    else:
                        d[pattern[i]] = word[i]
                else:
                    if d[pattern[i]] != word[i]:
                        return False
            
            return True
        
        result = []
        for i in range(0,len(words)):
            if is_matched(words[i], pattern) is True:
                result.append(words[i])
        
        return result